### PR TITLE
Add sesman-unlink-all-buffer

### PR DESCRIPTION
**Pull request by markgdawson**
Add an interactive utility function sesman-unlink-all-buffer to quickly unlinks all sessions linked to this buffer.

### CODE REVIEW

File: sesman.el [hunk 1]
```diff
@@ -473,6 +473,14 @@ buffer."
           (sesman--ask-for-link "Unlink: " links 'ask-all)))
   (run-hooks 'sesman-post-command-hook))
 
+;;;###autoload
+(defun sesman-unlink-all-context ()
+  "Unlink all sesman sessions in current context."
+  (interactive)
+  (mapc #'sesman--unlink
+        (sesman-current-links 'CIDER))
```
Comment (vspinu) [inline sesman.el]: Cider specific

File: sesman.el [hunk 2]
```diff
@@ -473,6 +473,14 @@ buffer."
           (sesman--ask-for-link "Unlink: " links 'ask-all)))
   (run-hooks 'sesman-post-command-hook))
 
+;;;###autoload
+(defun sesman-unlink-all-context ()
```
Comment (vspinu) [inline sesman.el]: Name should probably be just sesman-unlink-all.

**Review by vspinu — CHANGES_REQUESTED** [review]
`sesman-unlink` offers  [`*all*` option ](https://github.com/vspinu/sesman/blob/90974f9fcf944580de9c3698b5234fa57eddde1e/sesman.el#L301-L302) Why is this not enough?

**Comment by markgdawson** [comment]
@vspinu the `*all*` option at the moment can only be used through `completing-read`, which to my mind makes it awkward to bind to keys or to call from elisp snippets.

My use case is to call it non-interactively to force unlink everything so that the next link that I set is sure to be active. 

See here:

https://github.com/markgdawson/polybar-clj.el/blob/4297b91f7d1de68f00ea472d350baba4208869a0/polybar-clj.el#L254

It seemed like a useful function to share. And although I don't use it interactively (yet?),  I can imagine it being useful.

**Comment by vspinu** [comment]
How about adding `all` argument to `sesman-unlink` instead of a new function? Something like this:

```diff
   (sesman--link-session-interactively session nil nil))
 
 ;;;###autoload
-(defun sesman-unlink ()
-  "Break any of the previously created links."
-  (interactive)
+(defun sesman-unlink (&optional all)
+  "Break any of the previously created links.
+If ALL is non-nil, break all currently existing links."
+  (interactive "P")
   (let* ((system (sesman--system))
          (links (or (sesman-current-links system)
                     (user-error "No %s links found" system))))
     (mapc #'sesman--unlink
-          (sesman--ask-for-link "Unlink: " links 'ask-all)))
+          (if all
+              links
+            (sesman--ask-for-link "Unlink: " links 'ask-all))))
   (run-hooks 'sesman-post-command-hook))
 
 (declare-function sesman-browser "sesman-browser")
```

**Comment by markgdawson** [comment]
That works for me.

**Comment by markgdawson** [comment]
If also be happy with passing `links` into the function? Which could be more flexible?

**Comment by vspinu** [comment]
yep, makes sense indeed.

**Comment by markgdawson** [comment]
How about something like:
```diff
 ;;;###autoload
-(defun sesman-unlink ()
-  "Break any of the previously created links."
-  (interactive)
-  (let* ((system (sesman--system))
-         (links (or (sesman-current-links system)
-                    (user-error "No %s links found" system))))
-    (mapc #'sesman--unlink
-          (sesman--ask-for-link "Unlink: " links 'ask-all)))
+(defun sesman-unlink (&optional links)
+  "Break any of the previously created links or LINKS if provided."
+  (interactive (list (sesman--ask-for-link "Unlink: "
+                                           (or
+                                            (sesman-current-links (sesman--system))
+                                            (user-error "No %s links found" (sesman--system)))
+                                           'ask-all)))
+  (mapc #'sesman--unlink links)
   (run-hooks 'sesman-post-command-hook))
```

**Comment by markgdawson** [comment]
Something like:
```clojure
;;;###autoload
(defun sesman-unlink (&optional links)
  "Break any of the previously created links or break LINKS if non-nil."
  (interactive)
  (mapc #'sesman--unlink (or links
                             (sesman--ask-for-link "Unlink: "
                                                   (or
                                                    (sesman-current-links (sesman--system))
                                                    (user-error "No %s links found" (sesman--system)))
                                                   'ask-all)))
  (run-hooks 'sesman-post-command-hook))
```

**Comment by markgdawson** [comment]
On a related note, my understanding is that `--` in a name typically denoted something that is supposed to be internal to the package. So `sesman--system` might suggest it isn't for the public to use.

Should there be a `sesman-system` (without `--`) function for public consumption? So that something like the following can be used as part of a supported API?
```
(sesman-unlink (sesman-current-links (sesman-system)))
```

Or maybe I'm overthinking this?

File: sesman.el [hunk 3]
```diff
@@ -463,22 +463,15 @@ buffer."
   (sesman--link-session-interactively session nil nil))
 
 ;;;###autoload
-(defun sesman-unlink ()
-  "Break any of the previously created links."
+(defun sesman-unlink (&optional links)
```
Comment (vspinu) [inline sesman.el:466]: Can you make it work with C-u  interactively to mean all links?

**Comment by vspinu** [comment]
> Should there be a sesman-system (without --) 

`sesman-system` is already a variable which can be a function. This is why the `sesman--system`.  Can be named `sesman-get-system` or something, but I think sesman--system is so stable that it's very unlikely to change in the future.

**Comment by markgdawson** [comment]
Pushed an implementation which uses `current-prefix-args`, so that the behaviour of `(sesman-unlink)` is preserved.

**Comment by markgdawson** [comment]
> but I think `sesman--system` is so stable that it's very unlikely to change in the future.

This makes a lot of sense. Personally I'm more than happy to just use `sesman--system` in my stuff.

But maybe this is a communication thing? I had assumed that `sesman--system` was internal because of the `--`, and I was looking for something which was public and "safe to use". Other people might jump to the same conclusion?

There could be something like:
```
(defun sesman-get-system ()
  (sesman--system))`
 ```
 
Which could remove the confusion, and at the same time avoid breaking anything which currently uses `sesman--system`.

**Comment by vspinu** [comment]
Thanks!

Regarding sesman-get-system, it can be an alias. I will add that.

**Pull request merged**